File: day13/part1.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <queue>
#include <stdlib.h>

using namespace std;

int data[8][8] = {0};

vector<string> getLines() {
    ifstream file("input.txt");
    vector<string> lines;
    string line;

    while (getline(file, line))
        lines.push_back(line);

    return lines;
}

int nextPerson(queue<int> choices, int last) {
    int biggest = 0;
    
    for (int i = 0; i < choices.size(); i++) {
        int next = choices.front();
        choices.pop();

        int happiness = data[last][next] + data[next][last];
        happiness += nextPerson(choices, next);
        biggest = max(biggest, happiness);

        choices.push(next);
    }
    
    if (choices.size() == 0)
        biggest += data[last][0] + data[0][last];
    
    return biggest;
}

int main () {
    vector<string> lines = getLines();
    
    map<string, int> people = {
        {"Alice", 0},
        {"Bob", 1},
        {"Carol", 2},
        {"David", 3},
        {"Eric", 4},
        {"Frank", 5},
        {"George", 6},
        {"Mallory", 7}
    };
    queue<int> choices({1, 2, 3, 4, 5, 6, 7});
    
    for(vector<string>::iterator line = lines.begin(); line != lines.end(); ++line) {
        istringstream iss(*line);
        string person, otherPerson, action, garbage;
        int happiness;
        
        iss >> person >> garbage >> action >> happiness >> garbage >> garbage >> garbage >> garbage >> garbage >> garbage >> otherPerson;
        otherPerson.pop_back();
        
        if (action == "lose") happiness *= -1;
        data[people[person]][people[otherPerson]] = happiness;
    }
    
    cout << nextPerson(choices, 0) << endl;
}

File: 2015/ruby/day13/part1.rb
#!/usr/bin/ruby

def sit(sat, happiness)
    max = 0

    for i in 0..7
        next if sat.include?(i)
        max = [sit(sat.push(i), happiness), max].max
        sat.pop()
    end

    if sat.size == 8
        max += happiness[sat[0]][sat[7]] + happiness[sat[7]][sat[0]]
        for i in 1..7
            max += happiness[sat[i]][sat[i - 1]] + happiness[sat[i - 1]][sat[i]]
        end
    end

    return max
end

begin
    File.open('input.txt', 'r') do |file|
        people = {
            'Alice' => 0,
            'Bob' => 1,
            'Carol' => 2,
            'David' => 3,
            'Eric' => 4,
            'Frank' => 5,
            'George' => 6,
            'Mallory' => 7
        }
        happiness = Array.new(8) { Array.new(8) }

        file.each_line do |line|
            sentence = line[0..-3].split(' ')
            happiness[people[sentence[0]]][people[sentence[10]]] = sentence[3].to_i

            if sentence[2] == 'lose'
                happiness[people[sentence[0]]][people[sentence[10]]] *= -1
            end
        end

        puts sit([], happiness)
    end
rescue SystemCallError
    puts 'input.txt not found.'
end
